code/modules/movespeed/_movespeed_modifier.dm

```diff
@@ -0,0 +1,257 @@
+/**
+  * Movespeed modification datums.
+  */
+
+/datum/movespeed_modifier
+	/// Whether or not this is a variable modifier. Variable modifiers can NOT be ever auto-cached. ONLY CHECKED VIA INITIAL(), EFFECTIVELY READ ONLY (and for very good reason)
+	var/variable = FALSE
+
+	/// Unique ID. You can never have different modifications with the same ID
+	var/id = "ERROR"
+
+	/// Higher ones override lower priorities. This is NOT used for ID, ID must be unique, if it isn't unique the newer one overwrites automatically if overriding.
+	var/priority = 0
+	var/flags = NONE
+
+	/// Multiplicative slowdown
+	var/multiplicative_slowdown = 0
+
+	/// Movetypes this applies to
+	var/movetypes = ALL
+
+	/// Movetypes this never applies to
+	var/blacklisted_movetypes = NONE
+
+	/// Other modification datums this conflicts with.
+	var/conflicts_with
+
+/*! How move speed for mobs works
+
+Move speed is now calculated by using a list of movespeed modifiers, which is a list itself (to avoid datum overhead)
+
+This gives us the ability to have multiple sources of movespeed, reliabily keep them applied and remove them when they should be
+
+THey can have unique sources and a bunch of extra fancy flags that control behaviour
+
+Previously trying to update move speed was a shot in the dark that usually meant mobs got stuck going faster or slower
+
+This list takes the following format
+
+```Current movespeed modification list format:
+		list(
+			id = list(
+				priority,
+				flags,
+				legacy slowdown/speedup amount,
+				movetype_flags
+			)
+		)
+```
+
+WHen update movespeed is called, the list of items is iterated, according to flags priority and a bunch of conditions
+this spits out a final calculated value which is used as a modifer to last_move + modifier for calculating when a mob
+can next move
+
+Key procs
+* [add_movespeed_modifier](mob.html#proc/add_movespeed_modifier)
+* [remove_movespeed_modifier](mob.html#proc/remove_movespeed_modifier)
+* [has_movespeed_modifier](mob.html#proc/has_movespeed_modifier)
+* [update_movespeed](mob.html#proc/update_movespeed)
+*/
+
+//ANY ADD/REMOVE DONE IN UPDATE_MOVESPEED MUST HAVE THE UPDATE ARGUMENT SET AS FALSE!
+
+GLOBAL_LIST_EMPTY(movespeed_modification_cache)
+
+/// Grabs a STATIC MODIFIER datum from cache. YOU MUST NEVER EDIT THESE DATUMS, OR IT WILL AFFECT ANYTHING ELSE USING IT TOO!
+/proc/get_cached_movespeed_modifier(modtype)
+	if(!ispath(modtype, /datum/movespeed_modifier))
+		CRASH("[modtype] is not a movespeed modification typepath.")
+	var/datum/movespeed_modifier/M = modtype
+	if(initial(M.variable))
+		CRASH("[modtype] is a variable modifier, and can never be cached.")
+	return GLOB.movespeed_modification_cache[modtype] || ((GLOB.movespeed_modification_cache[modtype] = new modtype))
```

Comment:
Why the double wrapping?

Reply:
Micro optimization I learned when doing getflaticon, this **seems** to be faster than a var set behind an if block

Reply:
I'm sorry ***what***

I'll need to see profile data for that one

Reply:
I hate you both

Reply:
![image](https://user-images.githubusercontent.com/2003111/72308658-1fef4900-363a-11ea-890d-1bd8e94f95e7.png)
![image](https://user-images.githubusercontent.com/2003111/72308672-24b3fd00-363a-11ea-964f-e0b35d5682fa.png)

aight bet gamers